README.md
# http-echo-server

Forked from [watson/http-echo-server](https://github.com/watson/http-echo-server/)

Will accept any TCP connection and echo back an HTTP response with the
entire content of the incoming TCP connection.

The server makes no attempt to understand the incoming HTTP request
hence it doesn't know when the request is completed and therefore just
terminates the TCP connection 2 seconds after the first data packet.

[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](https://github.com/standard/standard)

## Installation

To set up a simple echo-server on Heroku just click this button:

[![Deploy](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy)

*Note that the Heroku routing stack will proxy the incoming request and
add custom HTTP headers.*

Alternatively, to start it locally just run (only supported on newer
versions of npm):

```
npx http-echo-server
```

Or if you wish to install the module globally:

```
npm install http-echo-server -g
```

## Example usage

Just curl the URL of the app:

```
curl http://<heroku-app-name>.herokuapp.com
```

Alternatively - if installed locally - you can start the server using
the command `http-echo-server`, take note of the port it starts on and
then curl it:

```
curl http://localhost:<port>
```

## Setting listening port

To set the http port, either supply the port as an argument to the
`http-echo-server` executable:

```
http-echo-server 3005
```

Or use the `PORT` environment variable:

```
export PORT=3005
http-echo-server
```

## Setting response delay

To set the response delay (in milliseconds), either supply the port as the 2nd argument to the
`http-echo-server` executable:

```
http-echo-server 3005 250
```

Or use the `PORT` environment variable:

```
export DELAY=250
http-echo-server
```

## License

[MIT](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.2.0](https://github.com/Mualig/http-echo-server/compare/v2.1.1...v2.2.0) (2020-12-16)


### Features

* **delay:** added configurable delay ([9f4439d](https://github.com/Mualig/http-echo-server/commit/9f4439d48e6cf2bf07a0742921e0c5746dfa7365)), closes [watson/http-echo-server#5](https://github.com/watson/http-echo-server/issues/5)
* **logs:** added timestamp to logs ([a750f37](https://github.com/Mualig/http-echo-server/commit/a750f378d48151c7c6607090ef52685d0cb86ff0))


test.js
'use strict'

const assert = require('assert')
const net = require('net')

const server = require('./')

server.on('listening', function () {
  const port = server.address().port
  const client = net.createConnection({ port: port })
  let result = ''
  client.on('data', function (chunk) {
    result += chunk
  })
  client.on('end', function () {
    const lines = result.split('\r\n')
    assert.strictEqual(lines.length, 7)
    assert.strictEqual(lines.shift(), 'HTTP/1.1 200 OK')
    assert.strictEqual(lines.shift().slice(0, 6), 'Date: ')
    assert.strictEqual(lines.shift(), 'Connection: close')
    assert.strictEqual(lines.shift(), 'Content-Type: text/plain')
    assert.strictEqual(lines.shift(), 'Access-Control-Allow-Origin: *')
    assert.strictEqual(lines.shift(), '')
    assert.strictEqual(lines.shift(), 'hello world')
    server.close()
  })
  client.end('hello world')
})


src/test.js
'use strict'

const assert = require('assert')
const net = require('net')

const server = require('./index')

server.on('listening', function () {
  const port = server.address().port
  const client = net.createConnection({ port: port })
  let result = ''
  client.on('data', function (chunk) {
    result += chunk
  })
  client.on('end', function () {
    const lines = result.split('\r\n')
    assert.strictEqual(lines.length, 7)
    assert.strictEqual(lines.shift(), 'HTTP/1.1 200 OK')
    assert.strictEqual(lines.shift().slice(0, 6), 'Date: ')
    assert.strictEqual(lines.shift(), 'Connection: close')
    assert.strictEqual(lines.shift(), 'Content-Type: text/plain')
    assert.strictEqual(lines.shift(), 'Access-Control-Allow-Origin: *')
    assert.strictEqual(lines.shift(), '')
    assert.strictEqual(lines.shift(), 'hello world')
    server.close()
  })
  client.end('hello world')
})


server.js
'use strict'

var getPort = require('get-port')
var server = require('net').createServer()

var cid = 0

module.exports = server // for testing

onEmit(server, {ignore: ['connection', 'listening', 'error']}, function (eventName) {
  console.log('[server] event:', eventName)
})

server.on('connection', function (c) {
  var gotData = false
  var _cid = ++cid

  console.log('[server] event: connection (socket#%d)', _cid)

  onEmit(c, {ignore: ['lookup', 'error']}, function (eventName) {
    console.log('[socket#%d] event:', _cid, eventName)
  })

  c.on('lookup', function (err, address, family) {
    if (err) {
      console.log('[socket#%d] event: lookup (error: %s)', _cid, err.message)
    } else {
      console.log('[socket#%d] event: lookup (address: %s, family: %s)', _cid, address, family)
    }
  })

  c.on('data', function (chunk) {
    console.log('--> ' + chunk.toString().split('\n').join('\n--> '))
    if (!gotData) {
      gotData = true
      c.write('HTTP/1.1 200 OK\r\n')
      c.write('Date: ' + (new Date()).toString() + '\r\n')
      c.write('Connection: close\r\n')
      c.write('Content-Type: text/plain\r\n')
      c.write('\r\n')
      setTimeout(function () {
        c.end()
      }, 2000)
    }
    c.write(chunk.toString())
  })

  c.on('error', function (err) {
    console.log('[socket#%d] event: error (msg: %s)', _cid, err.message)
  })
})

server.on('listening', function () {
  var port = server.address().port
  console.log('[server] event: listening (port: %d)', port)
})

server.on('error', function (err) {
  console.log('[server] event: error (msg: %s)', err.message)
})

if (process.env.PORT) {
  server.listen(process.env.PORT)
} else {
  getPort({port: 3000}).then(function (port) {
    server.listen(port)
  })
}

function onEmit (emitter, opts, cb) {
  var emitFn = emitter.emit
  emitter.emit = function (eventName) {
    if (opts.ignore.indexOf(eventName) === -1) cb.apply(null, arguments)
    return emitFn.apply(emitter, arguments)
  }
}


src/index.js
#!/usr/bin/env node
'use strict'

const getPort = require('get-port')
const server = require('net').createServer()

let cid = 0
let timeout = 2000

module.exports = server // for testing

onEmit(server, { ignore: ['connection', 'listening', 'error'] }, function (eventName) {
  console.log(`[${new Date().toISOString()}][server] event:`, eventName)
})

server.on('connection', function (socket) {
  let gotData = false
  const _cid = ++cid

  console.log(`[${new Date().toISOString()}][server] event: connection (socket#${_cid})`)

  onEmit(socket, { ignore: ['lookup', 'error'] }, function (eventName) {
    console.log(`[${new Date().toISOString()}][socket#${_cid}] event:`, eventName)
  })

  socket.on('lookup', function (err, address, family) {
    if (err) {
      console.log(`[${new Date().toISOString()}][socket#${_cid}] event: lookup (error: ${err.message})`)
    } else {
      console.log(`[${new Date().toISOString()}][socket#${_cid}] event: lookup (address: ${address}, family: ${family})`)
    }
  })

  socket.on('data', function (chunk) {
    console.log('--> ' + chunk.toString().split('\n').join('\n--> '))
    if (!gotData) {
      gotData = true
      socket.write('HTTP/1.1 200 OK\r\n')
      socket.write('Date: ' + (new Date()).toString() + '\r\n')
      socket.write('Connection: close\r\n')
      socket.write('Content-Type: text/plain\r\n')
      socket.write('Access-Control-Allow-Origin: *\r\n')
      socket.write('\r\n')
      setTimeout(function () {
        socket.end()
      }, timeout)
    }
    socket.write(chunk.toString())
  })

  socket.on('error', function (err) {
    console.log(`[${new Date().toISOString()}][socket#${_cid}] event: error (msg: ${err.message})`)
  })
})

server.on('listening', function () {
  const port1 = server.address().port
  console.log(`[${new Date().toISOString()}][server] event: listening (port: ${port1})`)
})

server.on('error', function (err) {
  console.log(`[${new Date().toISOString()}][server] event: error (msg: ${err.message})`)
})

const port = process.argv[2] || process.env.PORT

if (port) {
  server.listen(port)
} else {
  getPort({ port: 3000 }).then(function (port2) {
    server.listen(port2)
  })
}

const delay = process.argv[3] || process.env.DELAY

if (delay) {
  timeout = delay
}

console.log(`[${new Date().toISOString()}][server] delay: ${timeout}ms`)

function onEmit (emitter, opts, cb) {
  const emitFn = emitter.emit
  emitter.emit = function (eventName) {
    if (opts.ignore.indexOf(eventName) === -1) cb.apply(null, arguments)
    return emitFn.apply(emitter, arguments)
  }
}
